htlcswitch/mailbox.go

```diff
@@ -402,3 +402,174 @@ func (m *memoryMailBox) MessageOutBox() chan lnwire.Message {
+	liveIndex map[lnwire.ShortChannelID]lnwire.ChannelID
+
+	// TODO(conner): add another pair of indexes:
+	//   chan_id -> short_chan_id
+	//   short_chan_id -> mailbox
+	// so that Deliver can lookup mailbox directly once live,
+	// but still queriable by channel_id.
+
+	// unclaimedPackets maps a live short chan id to queue of packets if no
+	// mailbox has been created.
+	unclaimedPackets map[lnwire.ShortChannelID][]*htlcPacket
+}
+
+// newMailOrchestrator initializes a fresh mailOrchestrator.
+func newMailOrchestrator() *mailOrchestrator {
+	return &mailOrchestrator{
+		mailboxes:        make(map[lnwire.ChannelID]MailBox),
+		liveIndex:        make(map[lnwire.ShortChannelID]lnwire.ChannelID),
+		unclaimedPackets: make(map[lnwire.ShortChannelID][]*htlcPacket),
+	}
+}
+
+// Stop instructs the orchestrator to stop all active mailboxes.
+func (mo *mailOrchestrator) Stop() {
+	for _, mailbox := range mo.mailboxes {
+		mailbox.Stop()
+	}
+}
+
+// GetOrCreateMailBox returns an existing mailbox belonging to `chanID`, or
+// creates and returns a new mailbox if none is found.
+func (mo *mailOrchestrator) GetOrCreateMailBox(chanID lnwire.ChannelID) MailBox {
+	// First, try lookup the mailbox directly using only the shared mutex.
+	mo.mu.RLock()
+	mailbox, ok := mo.mailboxes[chanID]
+	if ok {
+		mo.mu.RUnlock()
+		return mailbox
+	}
+	mo.mu.RUnlock()
+
+	// Otherwise, we will try again with exclusive lock, creating a mailbox
+	// if one still has not been created.
+	mo.mu.Lock()
+	mailbox = mo.exclusiveGetOrCreateMailBox(chanID)
+	mo.mu.Unlock()
+
+	return mailbox
+}
+
+// exclusiveGetOrCreateMailBox checks for the existence of a mailbox for the
+// given channel id. If none is found, a new one is creates, started, and
+// recorded.
+//
+// NOTE: This method MUST be invoked with the mailOrchestrator's exclusive lock.
+func (mo *mailOrchestrator) exclusiveGetOrCreateMailBox(
+	chanID lnwire.ChannelID) MailBox {
+
+	mailbox, ok := mo.mailboxes[chanID]
+	if !ok {
+		mailbox = newMemoryMailBox()
+		mailbox.Start()
+		mo.mailboxes[chanID] = mailbox
+	}
+
+	return mailbox
+}
+
+// BindLiveShortChanID registers that messages bound for a particular short
+// channel id should be forwarded to the mailbox corresponding to the given
+// channel id. This method also checks to see if there are any unclaimed
+// packets for this short_chan_id. If any are found, they are delivered to the
+// mailbox and removed (marked as claimed).
+func (mo *mailOrchestrator) BindLiveShortChanID(mailbox MailBox,
+	cid lnwire.ChannelID, sid lnwire.ShortChannelID) {
+
+	mo.mu.Lock()
+	// Update the mapping from short channel id to mailbox's channel id.
+	mo.liveIndex[sid] = cid
+
+	// Retrieve any unclaimed packets destined for this mailbox.
+	pkts, _ := mo.unclaimedPackets[sid]
+	delete(mo.unclaimedPackets, sid)
+	mo.mu.Unlock()
+
+	// Deliver the unclaimed packets.
+	for _, pkt := range pkts {
+		mailbox.AddPacket(pkt)
+	}
```

Comment:
Can we also delete the set of unclaimed packages at this point?

Reply:
packets* 

But now see that they're deleted above, then fed into the mail box to be consumed by the link.